### PR TITLE
fix evaluator for language model

### PR DESCRIPTION
**Pull request by L1aoXingyu**
这个 PR 为语言模型增加了 evaluator，同时修复了一些问题

### CODE REVIEW

File: libai/data/datasets/t5_dataset.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def __getitem__(self, idx):
             encoder_attn_mask=DistTensorData(encoder_padding_mask),
             decoder_attn_mask=DistTensorData(decoder_padding_mask),
             encoder_decoder_attn_mask=DistTensorData(encoder_decoder_padding_mask),
-            labels=DistTensorData(labels, placement_idx=-1),
```
Comment (L1aoXingyu) [inline libai/data/datasets/t5_dataset.py:131]: 这里还是改回 lm_labels 比较符合 nlp 的语义

File: libai/engine/default.py [hunk 2]
```diff
@@ -639,7 +639,7 @@ def test(cls, cfg, test_loaders, model, evaluator=None):
         for idx, data_loader in enumerate(test_loaders):
             # When evaluators are passed in as arguments,
             # implicitly assume that evaluators can be created before data_loader.
-            dataset_name = getattr(data_loader.dataset, "dataset_name", "UndefinedDataset")
```
Comment (L1aoXingyu) [inline libai/engine/default.py:642]: 发现这里在 ConcatDataset 的时候其实也取不到 dataset_name，其实可以简单的用 class name 来替代这里的 dataset_name，这样也不用每个 dataset 都需要定义一个 dataset_name attribute

File: libai/evaluation/evaluator.py [hunk 3]
```diff
@@ -181,7 +181,13 @@ def inference_on_dataset(
 
             # get valid sample
             valid_data = {key: dist.ttol(value)[:valid_sample] for key, value in data.items()}
-            valid_outputs = {key: dist.ttol(value)[:valid_sample] for key, value in outputs.items()}
+            valid_outputs = {}
+            for key, value in outputs.items():
+                valid_outputs[key] = dist.ttol(
+                    value, ranks=[0] if value.placement.ranks.ndim == 1 else [[0]]
```
Comment (L1aoXingyu) [inline libai/evaluation/evaluator.py:187]: 这里是解决流水并行的时候，将结果放到 rank0 上

File: libai/evaluation/evaluator.py [hunk 4]
```diff
@@ -181,7 +181,13 @@ def inference_on_dataset(
...
+                )
+                if valid_outputs[key].ndim > 1:
```
Comment (L1aoXingyu) [inline libai/evaluation/evaluator.py:189]: 这里是考虑到 output 是 scalar 的情况，先采用这种方式绕过去

在 nlp 里面计算 ppl(困惑度) 是通过 loss 进行计算的，所以会返回一个 scalar

File: tools/train_cv.py [hunk 5]
```diff
@@ -19,9 +19,16 @@
 sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), os.path.pardir)))
```
Comment (L1aoXingyu) [inline tools/train_cv.py:19]: 因为 cv 和 nlp 不共享 evaluator，所以构建了两个 train_net.py 分别支持 cv 和 nlp 的场景

File: libai/models/t5_model.py [hunk 6]
```diff
@@ -368,8 +368,8 @@ def forward(
             encoder_decoder_attn_mask,
         )
 
-        if self.training and labels is not None:
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:371]: 这个类是 `T5ForPreTraining`，一般做下游任务只会把 T5 拆出来用，而不会使用这个类，所以这个类的作用就是做 pretraining，而在 pretraining 中，evaluation 也需要算 loss 来计算 ppl，所以这里不用判断是不是 training mode，主要提供了 labels 就计算对应的 loss

File: libai/evaluation/evaluator.py (hunk 4 above)
Comment (rentainhe) [reply libai/evaluation/evaluator.py:189]: 感觉写成`valid_outputs[key] = value[:valid_sample]`更好理解一些

File: tools/train_cv.py (hunk 5 above)
Comment (rentainhe) [reply tools/train_cv.py:19]: 这样是不是在启动的时候需要指定`.py`了

Comment (CPFLAME) [reply tools/train_cv.py:19]: 我想想  这里我们可以在config里面增加一个evaluator字段,  然后用lazycall的方式调用,  这样可以共用一个train_net.py

Comment (rentainhe) [reply tools/train_cv.py:19]: 我也觉得可能作为config传入会更合理一些

File: configs/common/train.py [hunk 7]
```diff
@@ -31,11 +32,16 @@
 
     checkpointer=dict(period=5000, max_to_keep=100),  # options for PeriodicCheckpointer
 
+    # options for evaluation
+    evaluation=dict(
```
Comment (L1aoXingyu) [inline configs/common/train.py:36]: 使用 lazyCall 的方式调用 evaluator，增加一个开关可以关掉 evaluation

File: libai/engine/default.py (hunk 2 above)
Comment (L1aoXingyu) [reply libai/engine/default.py:642]: 好的，应该就 cifar 和 imagenet 加过，nlp 的 dataset 还没加过

File: libai/engine/hooks.py [hunk 8]
```diff
@@ -236,33 +236,33 @@ def _best_checking(self):
...
-                    f"Given val metric {self._val_metric} does not seem to be computed/stored."
+                    f"Given val metric {self._val_metric} does not seem to be computed/stored. "
                     "Will not be checkpointed based on it."
                 )
             else:
                 latest_metric, metric_iter = metric_tuple
 
-            if self.best_metric is None:
-                if self._update_best(latest_metric, metric_iter):
+                if self.best_metric is None:
```
Comment (L1aoXingyu) [inline libai/engine/hooks.py:245]: 这里 fix 没有 metrics_tuple 时执行的问题

**Review by rentainhe — APPROVED** [review]

File: tools/train.sh [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 NODE=${NODE:-1}
 NODE_RANK=${NODE_RANK:-0}
 ADDR=${ADDR:-127.0.0.1}
 PORT=${PORT:-12345}
 
 python3 -m oneflow.distributed.launch \
 --nproc_per_node $GPUS --nnodes $NODE --node_rank $NODE_RANK --master_addr $ADDR --master_port $PORT \
-tools/train_net.py \
---config-file $CONFIG ${@:3}
+$FILE --config-file $CONFIG ${@:4}
```
Comment (L1aoXingyu) [inline tools/train.sh:13]: 在这里把默认用 `tools/train_net.py` 换成一个参数传入，这样 project 下面写的训练文件也可以复用的 sh

Comment (rentainhe) [reply tools/train.sh:13]: nice，但是我一般可能习惯在projects下也单独写一个`train_net.py`，貌似d2延伸出去的projects也是类似的

Comment (L1aoXingyu) [reply tools/train.sh:13]: 改完之后的 sh 可以直接调用 project 下面的 train_net 就想这样 `bash tools/train.sh project/xxx/train_net.py project/xxx/config.py 1 train.train_iter=10`

File: libai/evaluation/evaluator.py [hunk 10]
```diff
@@ -181,7 +181,13 @@ def inference_on_dataset(
...
             # get valid sample
             valid_data = {key: dist.ttol(value)[:valid_sample] for key, value in data.items()}
-            valid_outputs = {key: dist.ttol(value)[:valid_sample] for key, value in outputs.items()}
+            valid_outputs = {}
+            for key, value in outputs.items():
+                value = dist.ttol(value, ranks=[0] if value.placement.ranks.ndim == 1 else [[0]])
+                if value.ndim > 1:
+                    valid_outputs[key] = value[:valid_sample]  # Slice if it's batched output
+                else:
+                    valid_outputs[key] = value
```
Comment (L1aoXingyu) [inline libai/evaluation/evaluator.py:190]: 这一块程鹏再看看有没有更好的方案

Comment (CPFLAME) [reply libai/evaluation/evaluator.py:190]: 这里没什么更好的方案了感觉

**Review by CPFLAME — APPROVED** [review]
